### PR TITLE
in-memory-region: fix the inconsistency caused by updating region range

### LINKED ISSUE #17483: In-memory Engine: `iter_overlapped_regions` panics no entry found for key
## Bug Report

Running a custom sysbench workload, TiKV panics with the following stacktrace.

```log
[2024/09/01 14:15:38.837 +00:00] [FATAL] [lib.rs:478] ["no entry found for key"] [backtrace="   0: tikv_util::set_panic_hook::{{closure}}
             at home/stn/tikv/components/tikv_util/src/lib.rs:477:18
   1: <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call
             at home/stn/.rustup/toolchains/nightly-2023-12-28-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/boxed.rs:2029:9
      std::panicking::rust_panic_with_hook
             at home/stn/.rustup/toolchains/nightly-2023-12-28-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panicking.rs:783:13
   2: std::panicking::begin_panic_handler::{{closure}}
             at home/stn/.rustup/toolchains/nightly-2023-12-28-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panicking.rs:657:13
   3: std::sys_common::backtrace::__rust_end_short_backtrace
             at home/stn/.rustup/toolchains/nightly-2023-12-28-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/sys_common/backtrace.rs:171:18
   4: rust_begin_unwind
             at home/stn/.rustup/toolchains/nightly-2023-12-28-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panicking.rs:645:5
   5: core::panicking::panic_fmt
             at home/stn/.rustup/toolchains/nightly-2023-12-28-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/panicking.rs:72:14
   6: core::panicking::panic_display
             at home/stn/.rustup/toolchains/nightly-2023-12-28-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/panicking.rs:196:5
      core::panicking::panic_str
             at home/stn/.rustup/toolchains/nightly-2023-12-28-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/panicking.rs:171:5
      core::option::expect_failed
             at home/stn/.rustup/toolchains/nightly-2023-12-28-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/option.rs:1980:5
   7: core::option::Option<T>::expect
             at home/stn/.rustup/toolchains/nightly-2023-12-28-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/option.rs:894:21
      <std::collections::hash::map::HashMap<K,V,S> as core::ops::index::Index<&Q>>::index
             at home/stn/.rustup/toolchains/nightly-2023-12-28-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/collections/hash/map.rs:1338:23
      range_cache_memory_engine::range_manager::RegionManager::iter_overlapped_regions
             at home/stn/tikv/components/range_cache_memory_engine/src/range_manager.rs:318:44
      range_cache_memory_engine::range_manager::RegionManager::check_overlap_with_region
             at home/stn/tikv/components/range_cache_memory_engine/src/range_manager.rs:418:9
      range_cache_memory_engine::range_manager::RegionManager::load_region
             at home/stn/tikv/components/range_cache_memory_engine/src/range_manager.rs:762:30
   8: range_cache_memory_engine::engine::RangeCacheMemoryEngine::prepare_for_apply
             at home/stn/tikv/components/range_cache_memory_engine/src/engine.rs:348:16
      <range_cache_memory_engine::write_batch::RangeCacheWriteBatch as engine_traits::write_batch::WriteBatch>::prepare_for_region
             at home/stn/tikv/components/range_cache_memory_engine/src/write_batch.rs:494:37
   9: <raftstore::coprocessor::read_write::write_batch::WriteBatchWrapper<WB> as engine_traits::write_batch::WriteBatch>::prepare_for_region
             at home/stn/tikv/components/raftstore/src/coprocessor/read_write/write_batch.rs:118:13
      raftstore::store::fsm::apply::ApplyContext<EK>::prepare_for
             at home/stn/tikv/components/raftstore/src/store/fsm/apply.rs:547:9
  10: raftstore::store::fsm::apply::ApplyDelegate<EK>::handle_raft_committed_entries
             at home/stn/tikv/components/raftstore/src/store/fsm/apply.rs:1134:9
  11: raftstore::store::fsm::apply::ApplyFsm<EK>::handle_apply
             at home/stn/tikv/components/raftstore/src/store/fsm/apply.rs:4087:9
  12: raftstore::store::fsm::apply::ApplyFsm<EK>::handle_tasks
             at home/stn/tikv/components/raftstore/src/store/fsm/apply.rs:4441:25
  13: <raftstore::store::fsm::apply::ApplyPoller<EK> as batch_system::batch::PollHandler<raftstore::store::fsm::apply::ApplyFsm<EK>,raftstore::store::fsm::apply::ControlFsm>>::handle_normal
             at home/stn/tikv/components/raftstore/src/store/fsm/apply.rs:4733:9
  14: batch_system::batch::Poller<N,C,Handler>::poll
             at home/stn/tikv/components/batch-system/src/batch.rs:422:27
  15: batch_system::batch::BatchSystem<N,C>::start_poller::{{closure}}
             at home/stn/tikv/components/batch-system/src/batch.rs:551:17
      <std::thread::Builder as tikv_util::sys::thread::StdThreadBuildWrapper>::spawn_wrapper::{{closure}}
             at home/stn/tikv/components/tikv_util/src/sys/thread.rs:438:13
      std::sys_common::backtrace::__rust_begin_short_backtrace
             at home/stn/.rustup/toolchains/nightly-2023-12-28-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/sys_common/backtrace.rs:155:18
  16: std::thread::Builder::spawn_unchecked_::{{closure}}::{{closure}}
             at home/stn/.rustup/toolchains/nightly-2023-12-28-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/thread/mod.rs:529:17
      <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
             at home/stn/.rustup/toolchains/nightly-2023-12-28-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/panic/unwind_safe.rs:272:9
      std::panicking::try::do_call
             at home/stn/.rustup/toolchains/nightly-2023-12-28-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panicking.rs:552:40
      std::panicking::try
             at home/stn/.rustup/toolchains/nightly-2023-12-28-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panicking.rs:516:19
      std::panic::catch_unwind
             at home/stn/.rustup/toolchains/nightly-2023-12-28-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panic.rs:142:14
      std::thread::Builder::spawn_unchecked_::{{closure}}
             at home/stn/.rustup/toolchains/nightly-2023-12-28-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/thread/mod.rs:528:30
      core::ops::function::FnOnce::call_once{{vtable.shim}}
             at home/stn/.rustup/toolchains/nightly-2023-12-28-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ops/function.rs:250:5
  17: <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
             at home/stn/.rustup/toolchains/nightly-2023-12-28-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/boxed.rs:2015:9
      <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
             at home/stn/.rustup/toolchains/nightly-2023-12-28-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/boxed.rs:2015:9
      std::sys::unix::thread::Thread::new::thread_start
             at home/stn/.rustup/toolchains/nightly-2023-12-28-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/sys/unix/thread.rs:108:17
  18: start_thread
  19: __clone3
"] [location=components/range_cache_memory_engine/src/range_manager.rs:318] [thread_name=apply-1] [thread_id=117]
```

### What version of TiKV are you using?

4c1bfe1fe27e99b71b500191fe973f9db9286cb4: A testing branch that merges https://github.com/tikv/tikv/pull/17296 and https://github.com/tikv/tikv/pull/17359

### Steps to reproduce

```
sysbench --mysql-host=$HOST --mysql-user=root --mysql-port=$PORT --threads=10  --db-driver=mysql --report-interval=1 --mysql-db=jobs --time=3600000  --number_workspaces=2 --jobs_per_org=10000 --select_percentage=10 many_delete_mvcc.lua run
```

### What did you expect?

No panic

### What did happened?

See above.

### LINKED ISSUE #17484: In-memory Engine: `on_snapshot_load_finished` assert fails
## Bug Report

Running a custom sysbench workload, TiKV panics with the following stacktrace.

```log
[2024/09/04 21:54:40.821 +00:00] [FATAL] [lib.rs:478] ["assertion failed: region.contains_range(meta.get_region())"] [backtrace="   0: tikv_util::set_panic_hook::{{closure}}
             at home/stn/tikv/components/tikv_util/src/lib.rs:477:18
   1: <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call
             at home/stn/.rustup/toolchains/nightly-2023-12-28-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/boxed.rs:2029:9
      std::panicking::rust_panic_with_hook
             at home/stn/.rustup/toolchains/nightly-2023-12-28-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panicking.rs:783:13
   2: std::panicking::begin_panic_handler::{{closure}}
             at home/stn/.rustup/toolchains/nightly-2023-12-28-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panicking.rs:649:13
   3: std::sys_common::backtrace::__rust_end_short_backtrace
             at home/stn/.rustup/toolchains/nightly-2023-12-28-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/sys_common/backtrace.rs:171:18
   4: rust_begin_unwind
             at home/stn/.rustup/toolchains/nightly-2023-12-28-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panicking.rs:645:5
   5: core::panicking::panic_fmt
             at home/stn/.rustup/toolchains/nightly-2023-12-28-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/panicking.rs:72:14
   6: core::panicking::panic
             at home/stn/.rustup/toolchains/nightly-2023-12-28-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/panicking.rs:144:5
   7: range_cache_memory_engine::range_manager::RegionManager::iter_overlapped_regions_mut
      range_cache_memory_engine::background::BackgroundRunnerCore::on_snapshot_load_finished
             at home/stn/tikv/components/range_cache_memory_engine/src/background.rs:507:13
      <range_cache_memory_engine::background::BackgroundRunner as tikv_util::worker::pool::Runnable>::run::{{closure}}
             at home/stn/tikv/components/range_cache_memory_engine/src/background.rs:1027:28
      yatp::task::future::RawTask<F>::poll
             at home/stn/.cargo/git/checkouts/yatp-e704b73c3ee279b6/793be4d/src/task/future.rs:59:9
   8: yatp::task::future::TaskCell::poll
             at home/stn/.cargo/git/checkouts/yatp-e704b73c3ee279b6/793be4d/src/task/future.rs:103:9
      <yatp::task::future::Runner as yatp::pool::runner::Runner>::handle
             at home/stn/.cargo/git/checkouts/yatp-e704b73c3ee279b6/793be4d/src/task/future.rs:387:20
   9: <tikv_util::yatp_pool::YatpPoolRunner<T> as yatp::pool::runner::Runner>::handle
             at home/stn/tikv/components/tikv_util/src/yatp_pool/mod.rs:199:24
  10: yatp::pool::worker::WorkerThread<T,R>::run
             at home/stn/.cargo/git/checkouts/yatp-e704b73c3ee279b6/793be4d/src/pool/worker.rs:48:13
  11: yatp::pool::builder::LazyBuilder<T>::build::{{closure}}
             at home/stn/.cargo/git/checkouts/yatp-e704b73c3ee279b6/793be4d/src/pool/builder.rs:114:25
      std::sys_common::backtrace::__rust_begin_short_backtrace
             at home/stn/.rustup/toolchains/nightly-2023-12-28-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/sys_common/backtrace.rs:155:18
  12: std::thread::Builder::spawn_unchecked_::{{closure}}::{{closure}}
             at home/stn/.rustup/toolchains/nightly-2023-12-28-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/thread/mod.rs:529:17
      <core::panic::unwind_safe::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
             at home/stn/.rustup/toolchains/nightly-2023-12-28-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/panic/unwind_safe.rs:272:9
      std::panicking::try::do_call
             at home/stn/.rustup/toolchains/nightly-2023-12-28-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panicking.rs:552:40
      std::panicking::try
             at home/stn/.rustup/toolchains/nightly-2023-12-28-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panicking.rs:516:19
      std::panic::catch_unwind
             at home/stn/.rustup/toolchains/nightly-2023-12-28-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/panic.rs:142:14
      std::thread::Builder::spawn_unchecked_::{{closure}}
             at home/stn/.rustup/toolchains/nightly-2023-12-28-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/thread/mod.rs:528:30
      core::ops::function::FnOnce::call_once{{vtable.shim}}
             at home/stn/.rustup/toolchains/nightly-2023-12-28-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ops/function.rs:250:5
  13: <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
             at home/stn/.rustup/toolchains/nightly-2023-12-28-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/boxed.rs:2015:9
      <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once
             at home/stn/.rustup/toolchains/nightly-2023-12-28-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/alloc/src/boxed.rs:2015:9
      std::sys::unix::thread::Thread::new::thread_start
             at home/stn/.rustup/toolchains/nightly-2023-12-28-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/std/src/sys/unix/thread.rs:108:17
  14: start_thread
  15: __clone3
"] [location=components/range_cache_memory_engine/src/background.rs:509] [thread_name=background-range-load-worker-0] [thread_id=32]
```

### What version of TiKV are you using?

e3522883c263d16cb0c3bda9fb2729f4facc4330:  A testing branch that merges https://github.com/tikv/tikv/pull/17296 and https://github.com/tikv/tikv/pull/17359

### Steps to reproduce

```
sysbench --mysql-host=$HOST --mysql-user=root --mysql-port=$PORT --threads=10  --db-driver=mysql --report-interval=1 --mysql-db=jobs --time=3600000  --number_workspaces=2 --jobs_per_org=10000 --select_percentage=10 many_delete_mvcc.lua run
```

### What did you expect?

No panic.

### What did happened?

See above.

### PR DESCRIPTION
**Pull request by glorv**
### What is changed and how it works?

Issue Number: Close #17483, close #17484

What's Changed:

```commit-message
The PR fix the bug by `amend_pending_region` that causes inconsistency
between regions and regions_by_range because it may change the region
range in metadata but not update the key of regions_by_range. This PR
change the original `amend_pending_region` by first remove the old
region and add a new one.
BTW, this PR also add a `Drop` implementation for `RegionManager` in
test environment to check the consistency between regions and
regions_by_range to find this kind of bug easier.
```

### Related changes

- [ ] PR to update `pingcap/docs`/`pingcap/docs-cn`:
- [ ] Need to cherry-pick to the release branch

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: components/range_cache_memory_engine/src/range_manager.rs [hunk 1]
```diff
@@ -122,18 +122,13 @@ impl CacheRegionMeta {
...
 
-    // update region info due to epoch version changes. This can only
-    // happen for pending region because otherwise we will always update
-    // the region epoch with ApplyObserver(for loading/active regions) or
-    // no need to update the epoch for evicting regions.
-    fn amend_pending_region(&mut self, region: &CacheRegion) -> bool {
+    // check whether we can replace the current outdated pending region with the new
+    // one.
+    fn can_be_updated_to(&self, region: &CacheRegion) -> bool {
         assert!(self.region.id == region.id && self.region.epoch_version < region.epoch_version);
```
Comment (overvenus) [inline components/range_cache_memory_engine/src/range_manager.rs:128]: Could you log region.id and epoch_version when assert fails?

File: components/range_cache_memory_engine/src/write_batch.rs [hunk 2]
```diff
@@ -954,4 +954,43 @@ mod tests {
             .unwrap();
         assert_eq!(snap2.get_value(b"zkk11").unwrap().unwrap(), &val1);
     }
+
+    #[test]
+    fn test_write_batch_update_outdated_pending_region() {
```
Comment (overvenus) [inline components/range_cache_memory_engine/src/write_batch.rs:959]: The current master branch also pass the test.

Comment (glorv) [reply components/range_cache_memory_engine/src/write_batch.rs:959]: Done, add a check for `regions_by_range`

**Review by overvenus — APPROVED** [review]

**Review by SpadeA-Tang — APPROVED** [review]

**Pull request merged**